lib/Sema/MiscDiagnostics.cpp

```diff
@@ -221,8 +221,25 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
       }
       
       // If we have an assignment expression, scout ahead for acceptable _'s.
-      if (auto *AE = dyn_cast<AssignExpr>(E))
-        markAcceptableDiscardExprs(AE->getDest());
+      if (auto *AE = dyn_cast<AssignExpr>(E)) {
+        if (auto destExpr = AE->getDest()) {
```

Comment:
There should always be a destination in AST, so you could simplify assert here to reduce nesting.